clippy_lints/src/methods/search_is_some.rs

```diff
@@ -146,3 +173,203 @@ pub(super) fn check<'tcx>(
         }
     }
 }
+
+struct ClosureSugg {
+    applicability: Applicability,
+    suggestion: String,
+}
+
+// Build suggestion gradually by handling closure arg specific usages,
+// such as explicit deref and borrowing cases.
+// Returns `None` if no such use cases have been triggered in closure body
+fn get_closure_suggestion<'tcx>(
+    cx: &LateContext<'_>,
+    search_arg: &'tcx hir::Expr<'_>,
+    closure_body: &hir::Body<'_>,
+) -> Option<ClosureSugg> {
+    let mut visitor = DerefDelegate {
+        cx,
+        closure_span: search_arg.span,
+        next_pos: search_arg.span.lo(),
+        suggestion_start: String::new(),
+        applicability: Applicability::MachineApplicable,
+    };
+
+    let fn_def_id = cx.tcx.hir().local_def_id(search_arg.hir_id);
+    cx.tcx.infer_ctxt().enter(|infcx| {
+        ExprUseVisitor::new(&mut visitor, &infcx, fn_def_id, cx.param_env, cx.typeck_results())
+            .consume_body(closure_body);
+    });
+
+    if visitor.suggestion_start.is_empty() {
+        None
+    } else {
+        Some(ClosureSugg {
+            applicability: visitor.applicability,
+            suggestion: visitor.finish(),
+        })
+    }
+}
+
+struct DerefDelegate<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    closure_span: Span,
+    next_pos: BytePos,
+    suggestion_start: String,
+    applicability: Applicability,
+}
+
+impl DerefDelegate<'_, 'tcx> {
+    pub fn finish(&mut self) -> String {
+        let end_span = Span::new(self.next_pos, self.closure_span.hi(), self.closure_span.ctxt());
+        let end_snip = snippet_with_applicability(self.cx, end_span, "..", &mut self.applicability);
+        format!("{}{}", self.suggestion_start, end_snip)
+    }
+
+    fn func_takes_arg_by_ref(&self, parent_expr: &'tcx hir::Expr<'_>, cmt_hir_id: HirId) -> bool {
+        if_chain! {
+            if let ExprKind::Call(func, call_args) = parent_expr.kind;
+            let typ = self.cx.typeck_results().expr_ty(func);
+            if let ty::FnDef(..) = typ.kind();
```

Comment:
And to make this function a bit more complicated, I believe we should also cover the cases of `ty::FnPtr(..)`, `ty::Closure(..)` and maybe `ty::Dynamic` as it looks like that can also be used for calls ([docs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.TyKind.html#variant.Dynamic)).

Warning regarding this change: I'm not totally sure if all of these cases are covered by `typ.fn_sig(self.cx.tcx).skip_binder().inputs()` in line 237 though

Reply:
I rewrote this function to be used with `MethodCall`, and removed this filter.

To be honest I can't remember why I put this filter on, and cannot find a reason to keep it, so let me know if you think it's a good idea to restrict again types here.